### PR TITLE
Do not log fatal while parsing

### PR DESCRIPTION
**Pull request by ibuildthecloud**
Signed-off-by: Darren Shepherd <darren@rancher.com>

### CODE REVIEW

**Comment by ibuildthecloud** [comment]
The causes badness if somebody uses libcompose in a server application and has a bad formatted file.  We would just `exit 1`, so yeah, that was bad.

**Comment by dnephin** [comment]
A couple unit tests for this case would be great

**Comment by ibuildthecloud** [comment]
@dnephin Updated and added a unit test.

File: project/merge.go [hunk 1]
```diff
@@ -37,7 +37,8 @@ func Merge(p *Project, bytes []byte) (map[string]*ServiceConfig, error) {
 	datas := make(rawServiceMap)
 	err := yaml.Unmarshal(bytes, &datas)
 	if err != nil {
-		logrus.Fatalf("Could not parse config for project %s : %v", p.Name, err)
+		logrus.Errorf("Could not parse config for project %s : %v", p.Name, err)
+		return nil, err
```
Comment (dnephin) [inline project/merge.go]: It seems like this `Merge` should be private?  Is it really part of the public interface? It's currently only called from one place in the same package, and from my reading seems like it shouldn't be public.

File: project/project.go [hunk 2]
```diff
@@ -123,7 +123,8 @@ func (p *Project) Load(bytes []byte) error {
 	configs := make(map[string]*ServiceConfig)
 	configs, err := Merge(p, bytes)
 	if err != nil {
-		log.Fatalf("Could not parse config for project %s : %v", p.Name, err)
+		log.Errorf("Could not parse config for project %s : %v", p.Name, err)
+		return err
```
Comment (dnephin) [inline project/project.go]: Merge just logged this error, do we need to log it a second time?

File: project/project_test.go [hunk 3]
```diff
@@ -19,3 +20,27 @@ func TestEventEquality(t *testing.T) {
...
+	}
+
+	p = NewProject(&Context{
+		ComposeBytes: []byte("not-garbage:\n  image: foo"),
+	})
+
+	err = p.Parse()
+	if err != nil {
+		t.Fatal(err)
+	}
```
Comment (dnephin) [inline project/project_test.go]: I always put the "success" case and "failure" case in separate tests so the output summary is more clear.  I'm not sure if this is a common practice in golang or not.

File: project/project_test.go [hunk 4]
```diff
@@ -19,3 +20,27 @@ func TestEventEquality(t *testing.T) {
 		t.Fatal("Events match")
 	}
 }
+
+func TestBadFile(t *testing.T) {
```
Comment (dnephin) [inline project/project_test.go]: Could this be `TestParseWithBadFile()` so it's clear which function is being tested?

**Comment by dnephin** [comment]
mostly looks good, just some really minor nits and questions

File: project/project.go (hunk 2 above)
Comment (vdemeester) [inline project/project.go]: Agree on that, it seems this one is not needed :wink:

File: project/project_test.go (hunk 3 above)
Comment (vdemeester) [inline project/project_test.go]: Same, I would split this test into `TestParseWithBadYaml` and `TestParseWithValidYaml`.

**Comment by ibuildthecloud** [comment]
@dnephin @vdemeester Updated based on comment.  Everything should be addressed.

**Comment by vdemeester** [comment]
LGTM 🐹

**Pull request merged**